src/core/lib/iomgr/ev_apple.cc

```diff
@@ -0,0 +1,252 @@
+  grpc_apple_register_write_stream_impl(write_stream, dispatch_queue);
+}
+
+/// Drive the run loop in a global singleton thread until the global run loop is
+/// shutdown.
+void GlobalRunLoopFunc(void* arg) {
+  grpc_core::ReleasableMutexLock lock(&gGlobalRunLoopContext->mu);
+  gGlobalRunLoopContext->run_loop = CFRunLoopGetCurrent();
+  gGlobalRunLoopContext->init_cv.Signal();
+
+  while (!gGlobalRunLoopContext->shutdown) {
+    // CFRunLoopRun() will return immediately if no stream is registered on it.
+    // So we wait on a conditional variable until a stream is registered;
+    // otherwise we'll be running a spinning loop.
+    while (!gGlobalRunLoopContext->input_source_registered) {
+      gGlobalRunLoopContext->input_source_cv.Wait(&gGlobalRunLoopContext->mu);
+    }
+    gGlobalRunLoopContext->input_source_registered = false;
+    lock.Unlock();
+    CFRunLoopRun();
+    lock.Lock();
+  }
+  lock.Unlock();
+}
+
+// pollset implementation
+
+static void pollset_global_init(void) {
+  gGlobalRunLoopContext = new GlobalRunLoopContext;
+
+  grpc_apple_register_read_stream_impl =
+      grpc_apple_register_read_stream_run_loop;
+  grpc_apple_register_write_stream_impl =
+      grpc_apple_register_write_stream_run_loop;
+
+  grpc_core::MutexLock lock(&gGlobalRunLoopContext->mu);
+  gGlobalRunLoopThread =
+      new grpc_core::Thread("apple_ev", GlobalRunLoopFunc, nullptr);
+  gGlobalRunLoopThread->Start();
+  while (gGlobalRunLoopContext->run_loop == NULL)
+    gGlobalRunLoopContext->init_cv.Wait(&gGlobalRunLoopContext->mu);
+}
+
+static void pollset_global_shutdown(void) {
+  {
+    grpc_core::MutexLock lock(&gGlobalRunLoopContext->mu);
+    gGlobalRunLoopContext->shutdown = true;
+    CFRunLoopStop(gGlobalRunLoopContext->run_loop);
+  }
+  gGlobalRunLoopThread->Join();
+  delete gGlobalRunLoopThread;
+  delete gGlobalRunLoopContext;
+}
+
+static void pollset_init(grpc_pollset* pollset, gpr_mu** mu) {
+  GRPC_POLLING_TRACE("pollset init: %p", pollset);
+  GrpcApplePollset* apple_pollset = new (pollset) GrpcApplePollset();
+  *mu = apple_pollset->mu.get();
+}
+
+static void pollset_shutdown(grpc_pollset* pollset, grpc_closure* closure) {
+  GRPC_POLLING_TRACE("pollset shutdown: %p", pollset);
+  grpc_core::ExecCtx::Run(DEBUG_LOCATION, closure, GRPC_ERROR_NONE);
+}
+
+void pollset_destroy(grpc_pollset* pollset) {
+  GRPC_POLLING_TRACE("pollset destroy: %p", pollset);
+  GrpcApplePollset* apple_pollset =
+      reinterpret_cast<GrpcApplePollset*>(pollset);
+  apple_pollset->~GrpcApplePollset();
+}
+
+grpc_error* pollset_work(grpc_pollset* pollset, grpc_pollset_worker** worker,
+                         grpc_millis deadline) {
+  GRPC_POLLING_TRACE("pollset work: %p, deadline: %" PRIu64, pollset, deadline);
+  GrpcApplePollset* apple_pollset =
+      reinterpret_cast<GrpcApplePollset*>(pollset);
+  if (worker != nullptr) {
+    *worker = nullptr;
+  }
+  while (!apple_pollset->kicked) {
+    if (apple_pollset->cv.Wait(
```

Comment:
Why don't we need to grab the pollset mutex beforehand? If it is required to hold the mutex before calling pollset_work, this should be commented out.

Reply:
It is exactly because [it is required to hold the mutex before calling grpc_pollset_work](https://github.com/grpc/grpc/blob/b25682427c748cc4cc71bded16334010c6d1cafc/src/core/lib/iomgr/pollset.h#L70). I don't really understand what you mean by "this should be commented out"